Resume from checkpoint file

cdcapano opened the pull request:
This patch adds the ability to allow `pycbc_inference` to resume from a checkpoint file if it was killed while still running. This should allow for smoother use on a cluster, particularly for doing PP tests.

Instead of dumping results to the output file, `pycbc_inference` now writes to `{output_file}.checkpoint`. In addition, a backup file is kept, called `{output_file}.bkup`. At each checkpoint results are written to `.checkpoint` and `.backup`. After each write, both the checkpoint file and backup file are checked using `inference_hdf.check_integrity`; it is also required that both files have the same number of samples. If one file fails the integrity check, the good file is copied over. At the end of successful run, the `.checkpoint` file is renamed to the output, and the backup file is deleted.

If the job is killed part way, then, if `--restore-from-checkpoint` is turned on, samples will automatically be loaded from the checkpoint file (done by using the checkpoint file in place of the samples file... the difference from samples file is that new results are written to the checkpoint file, which is not true of the samples file), and the sampler will resume from the last point. Upon starting, if the checkpoint file fails integrity checks for any reason, then results will be loaded from the backup file instead, and the backup file will be copied to the checkpoint file. This should ensure that the job can always resume, even if it failed while in the midst of writing to file.

I've tested with `emcee_pt` and the normal test distribution. Marking as work in progress as I haven't tested with the other samplers yet.

cdcapano commented:
@soumide1102 @cmbiwer Ok, I've tested with all of the samplers.

cdcapano commented:
@cmbiwer @soumide1102 poke

duncan-brown commented:
Soumi asked me to review this, so adding my name to the reviewers.

duncan-brown commented:
@cdcapano lack of checkpointing is becoming a problem for getting p-p results out, so can we merge this if it just works with  ``emcee_pt`` and worry about other samplers later?

duncan-brown commented:
Soumi says I should read the PR before commenting...

cdcapano commented:
@duncan-brown Yeah, it should work with all of the samplers.

@cmbiwer In reply to your questions on slack regarding the effect of this on run time: I haven't run it on a large job yet, so I can't be certain about performance. On my small analytic runs I didn't notice any difference. I don't think it should have much of an effect. `check_integrity` runs fairly quickly (< 1s) regardless of file size, and it is only called at each checkpoint. So it should be a negligible cost over the entire run.

As to rewriting the command line in the file: it's a good point to save the most recent command line, but I don't want to lose the information about what was done prior to a resumption (in case someone changed the command line). So, I've updated the patch to add the current command line to the file's ``cmd`` attrs. This means that ``cmd`` is now a list of strings, although I made it so that ``InferenceFile.cmd`` just returns the last one, so as not to break anything else that might currently use that attribute. I've also added an attribute that saves the number of iterations in the file if/when the job resumes from checkpoint. This is so you can know what the command line used was when a resume happened (plus it's just good information to have).

@duncan-brown @soumide1102 You should be able to use this with the jobs that you've already started in your pp tests. Just add ".checkpoint" to the output file's name and add "--resume-from-checkpoint" to the command-line options used for `pycbc_inference` (not sure how easy that is to do in a dax). The re-starting it should just pick up where it left off.

cmbiwer commented:
> So, I've updated the patch to add the current command line to the file's cmd attrs. This means that cmd is now a list of strings

This is what I meant. Thanks.

duncan-brown commented:
OK, Soumi helped me get some test cases set up, so I'll get this reviewed by the end of today.

cdcapano commented:
@duncan-brown Pokity McPokeface

cdcapano commented:
@duncan-brown I noticed a bug in this: when I added the commit to write the command line to the checkpoint file, I forgot to also do that to the backup file. Fixed in the commit I just pushed.

duncan-brown commented:
These matchsticks will hold the wires in the socket for now, but a better solution is needed long term that plays nicely with workflow management. Opened https://github.com/ligo-cbc/pycbc/issues/2078 to keep track.

https://www.youtube.com/watch?v=5N0ATdDdwKg
